File: qna/src/main/java/com/example/rag/qna/adapter/output/question/QuestionPersistenceAdapter.java
package com.example.rag.qna.adapter.output.question;

import com.example.rag.qna.domain.Question;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;

@Component
@AllArgsConstructor
public class QuestionPersistenceAdapter implements QuestionPersistencePort {

    private final QuestionRepository questionRepository;

    @Override
    public void saveQuestion(Question q) {
        questionRepository.save(
                new QuestionEntity(q.getCategory(), q.getTitle(), q.getContent(), q.getUserId(), LocalDateTime.now(), LocalDateTime.now()));
    }

    @Override
    public void updateQuestion(Question question) {
        QuestionEntity entity = getQuestionEntity(question.getQuestionId());
        entity.setCategory(question.getCategory());
        entity.setTitle(question.getTitle());
        entity.setContent(question.getContent());
        entity.setUpdatedAt(question.getUpdatedAt());
    }

    @Override
    public void deleteQuestion(String questionId) {
        questionRepository.deleteById(getQuestionIdForMySQL(questionId));
    }

    @Override
    public Question findQuestionById(String questionId) {
        return toQuestion(getQuestionEntity(questionId));
    }

    @Override
    public List<Question> findQuestionsByUserId(String userId, int offset, int limit) {
        return questionRepository.findByUserId(userId, offset, limit)
                .stream()
                .map(QuestionPersistenceAdapter::toQuestion)
                .toList();
    }

    private QuestionEntity getQuestionEntity(String questionId) {
        return questionRepository.findById(getQuestionIdForMySQL(questionId)).orElseThrow();
    }

    private static Question toQuestion(QuestionEntity entity) {
        return new Question(
                getQuestionIdForTransparency(entity.getId()),
                entity.getCategory(),
                entity.getTitle(),
                entity.getContent(),
                entity.getUserId(),
                entity.getCreatedAt(),
                entity.getUpdatedAt()
        );
    }

    private static Long getQuestionIdForMySQL(String questionId) {
        return questionId == null ? null : Long.parseLong(questionId);
    }

    private static String getQuestionIdForTransparency(Long questionId) {
        return questionId + "";
    }
}

File: qna/src/main/kotlin/com/example/rag/qna/adapter/output/question/QuestionPersistenceAdapter.kt
package com.example.rag.qna.adapter.output.question

import com.example.rag.qna.adapter.output.DBTransparency.idForMySQL
import com.example.rag.qna.adapter.output.DBTransparency.idForTransparency
import com.example.rag.qna.domain.Question
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class QuestionPersistenceAdapter(private val questionRepository: QuestionRepository) : QuestionPersistencePort {
    private companion object QuestionMapper {
        fun Question.createEntity() = QuestionEntity(getCategory(), getTitle(), getContent(), getUserId(), LocalDateTime.now())
        fun QuestionEntity.toDomainModel() = Question(idForTransparency(id), category, title, content, userId, createdAt, updatedAt)
        fun QuestionEntity.updateWithDomain(domain: Question) {
            category = domain.getCategory()
            title = domain.getTitle()
            content = domain.getContent()
            updatedAt = domain.getUpdatedAt()!!
        }
    }

    override fun saveQuestion(domain: Question): Question = questionRepository.save(domain.createEntity()).toDomainModel()
    override fun updateQuestion(domain: Question): Unit = getQuestionEntity(domain.getQuestionId()!!).updateWithDomain(domain)
    override fun deleteQuestion(questionId: String): Unit = questionRepository.deleteById(idForMySQL(questionId))
    override fun findQuestionById(questionId: String): Question = getQuestionEntity(questionId).toDomainModel()
    override fun findQuestionsByUserId(userId: String, offset: Int, limit: Int) = getQuestions(userId, offset, limit)

    private fun getQuestionEntity(questionId: String): QuestionEntity = questionRepository.findById(idForMySQL(questionId)).orElseThrow()
    private fun getQuestions(userId: String, offset: Int, limit: Int) = questionRepository.findByUserId(userId, offset, limit)
                                                                                          .map { entity -> entity.toDomainModel() }
}